src/coreclr/jit/optimizer.cpp

```diff
@@ -6054,46 +6071,60 @@ void Compiler::optHoistLoopNest(unsigned lnum, LoopHoistContext* hoistCtxt)
     m_loopsConsidered++;
 #endif // LOOP_HOIST_STATS
 
-    optHoistThisLoop(lnum, hoistCtxt);
-
-    VNSet* hoistedInCurLoop = hoistCtxt->ExtractHoistedInCurLoop();
+    BasicBlockList* preHeadersOfChildLoops = nullptr;
+    BasicBlockList* firstPreHeader         = nullptr;
 
     if (optLoopTable[lnum].lpChild != BasicBlock::NOT_IN_LOOP)
     {
-        // Add the ones hoisted in "lnum" to "hoistedInParents" for any nested loops.
-        // TODO-Cleanup: we should have a set abstraction for loops.
-        if (hoistedInCurLoop != nullptr)
-        {
-            for (VNSet::KeyIterator keys = hoistedInCurLoop->Begin(); !keys.Equal(hoistedInCurLoop->End()); ++keys)
-            {
-#ifdef DEBUG
-                bool b;
-                assert(!hoistCtxt->m_hoistedInParentLoops.Lookup(keys.Get(), &b));
-#endif
-                hoistCtxt->m_hoistedInParentLoops.Set(keys.Get(), true);
-            }
-        }
+        BitVecTraits m_visitedTraits(fgBBNumMax * 2, this);
+        BitVec       m_visited(BitVecOps::MakeEmpty(&m_visitedTraits));
 
+        // Since loops in optLoopTable are arranged in reverse execution order, arrange
+        // them back in execution order before processing.
```

Comment:
This is misleading; there is no defined execution order for the child loops.

Also, why do we care about the ordering?

Reply:
This was done as part of adding `vnSetHoistedForSibling`. Turns out that we don't need it anymore, I can well not do the reordering. 

But here are my notes just to show the problem. Imagine we have a method that has following loop hierarchy. 

```
L00   
   L01
   L02
      L03 
      L04
   L05
      L06
```
Earlier, we would iterate from outer to inner, so we would visit parent loop if there is anything to hoist and then look into child loops. I am changing it to first look into child loops and then look into parent loop. The reason being that if we hoist out something from inner loop, by looking in parent *later* we can try to hoist that expression further out of parent loop as well. So far good.I was trying to use a cache where I can check if an expression with same VN has been already hoisted by a sibling loop (loop at same level, eg. L01, L02 and L05 is one sibling group; L03,L04 is another), then skip hoisting it. So I would imagine, I would first inspect L01, hoist stuff out of it. When I inspect L02 I can check in cache if it was something already hoisted and not worry hoisting it. However, I discovered that we don't even iterate over the child loops in execution order. We iterate over them in reverse execution order. Basically with my change of inner to outer, it would be like this: 

```
 L05 
 L06
 L02
 L04
 L03
 L01
```

The downside of reverse order is, if I hoist something out of L05,  I can't use cache during `L01` processing if something is already hoisted by sibling loop because the hoisted expression follows after L01 in execution order. 

Ideally, the processing order should be:
 
```
 L01
 L03
 L04
 L02
 L06
 L05
```

~If I don't fix it, not using cache is an alternative, but turns out cache plays a big role in determining the profitability. Every time we hoist an expression, we could no. of expressions hoisted. When calculating profitability, we relate no. of expressions hoisted vs. no. of registers we have and might reject if we exceed number of expressions hoisted for a loop.~ Turns out this is not much benefit and it is very rare to have multiple similar expressions that are all present in nested loops and are getting hoisted.

Reply:
When reverting the "reverse" flow, I noticed yet another example of https://github.com/dotnet/runtime/issues/66994.

Just because the preheader's `bbNum` are different, we see diffs.

![image](https://user-images.githubusercontent.com/12488060/172680652-19a4db5b-2306-4f8b-ae9f-b518bed1afd6.png)

https://www.diffchecker.com/WOEwzTNo